api/src/main/java/org/openmrs/api/StorageService.java

```diff
@@ -0,0 +1,94 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.api;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.stream.Stream;
+
+/**
+ * Storage service to persist data that does not fit well in DB.
+ * <p>
+ * Data objects are stored under generated or provided unique keys. Data and keys are immutable.
+ * 
+ * @since 2.8.0, 2.7.2, 2.6.16, 2.5.15
+ */
+public interface StorageService extends OpenmrsService {
+	
+	/**
+	 * Get InputStream to read data for the given key.
+	 *
+	 * @param key unique key
+	 * @return data
+	 * @throws IOException wrong key or IO error
+	 */
+	InputStream getData(String key) throws IOException;
+	
+	/**
+	 * Returns keys starting with the given prefix (the order depends on the implementation).
+	 * <p>
+	 * Remember to close the stream when done iterating.
+	 *
+	 * @param moduleId null for global or module id
+	 * @param prefix key prefix
+	 * @return stream of keys
+	 * @throws IOException IO error
+	 */
+	Stream<String> getKeys(String moduleId, String prefix) throws IOException;
```

Comment:
What is the use case for this?

Reply:
It's pretty common to list files in a directory, e.g. https://github.com/openmrs/openmrs-module-reporting/blob/master/api/src/main/java/org/openmrs/module/reporting/definition/library/implementerconfigured/BaseImplementerConfiguredDefinitionLibrary.java#L125

Reply:
Oh i see! 😊

Reply:
So would a key be a folder name? For instance `lucene` as we currently have in the application data directory?

Reply:
Key points to an actual file not a directory, but a key can include a directory e.g. lucene/data.dat

Reply:
Can you also give an example of how the `prefix` would differ from the key?

Reply:
I call it prefix, because it matches any keys starting with the prefix e.g. having 'abc/ab', 'abc/abc', 'ab', 'abc/abcd', 'abc/ac' and the prefix 'abc/' we would get back keys 'abc/ab', 'abc/abc', 'abc/abcd' but not 'ab' and 'abc/ac'.

Reply:
Changed the var to keyPrefix. Hope it is more clear now.